python/ray/data/_internal/execution/operators/map_operator_state.py

```diff
@@ -0,0 +1,506 @@
+        # Whether we have started the task submitter yet.
+        self._have_started_submitter = False
+
+        # Put the function def in the object store to avoid repeated serialization
+        # in case it's large (i.e., closure captures large objects).
+        self._transform_fn_ref = ray.put(transform_fn)
+
+        # The temporary block bundle used to accumulate inputs until they meet the
+        # min_rows_per_bundle requirement.
+        self._block_bundle: Optional[RefBundle] = None
+
+        # Execution state.
+        self._tasks: Dict[ObjectRef[Union[ObjectRefGenerator, Block]], _TaskState] = {}
+        self._tasks_by_output_order: Dict[int, _TaskState] = {}
+        self._next_task_index: int = 0
+        self._next_output_index: int = 0
+        self._obj_store_mem_alloc: int = 0
+        self._obj_store_mem_freed: int = 0
+        self._obj_store_mem_cur: int = 0
+        self._obj_store_mem_peak: int = 0
+
+    def add_input(self, bundle: RefBundle) -> None:
+        if not self._have_started_submitter:
+            # Start the task submitter on the first input.
+            self._task_submitter.start()
+            self._have_started_submitter = True
+
+        if self._min_rows_per_bundle is None:
+            self._create_task(bundle)
+            return
+
+        def get_num_rows(bundle: Optional[RefBundle]):
+            if bundle is None:
+                return 0
+            if bundle.num_rows() is None:
+                return float("inf")
+            return bundle.num_rows()
+
+        bundle_rows = get_num_rows(bundle)
+        acc_num_rows = get_num_rows(self._block_bundle) + bundle_rows
+        if acc_num_rows > self._min_rows_per_bundle:
+            if self._block_bundle:
+                if get_num_rows(self._block_bundle) > 0:
+                    self._create_task(self._block_bundle)
+                self._block_bundle = bundle
+            else:
+                self._create_task(bundle)
+        else:
+            # TODO(ekl) add a warning if we merge 10+ blocks per bundle.
+            self._block_bundle = merge_ref_bundles(self._block_bundle, bundle)
+
+    def inputs_done(self, input_index: int) -> None:
+        assert input_index == 0, "Map operator only supports one input."
+        if self._block_bundle:
+            self._create_task(self._block_bundle)
+            self._block_bundle = None
+        self._task_submitter.task_submission_done()
+
+    def work_completed(self, ref: ObjectRef[Union[ObjectRefGenerator, Block]]) -> None:
+        self._task_submitter.task_done(ref)
+        task: _TaskState = self._tasks.pop(ref)
+        if task.block_metadata_ref is not None:
+            # Non-dynamic block splitting path.
+            # TODO(Clark): Remove this special case once dynamic block splitting is
+            # supported for actors.
+            block_refs = [ref]
+            block_metas = [ray.get(task.block_metadata_ref)]
```

Comment:
I wasn't able to isolate this task-actor submitter difference to the abstractions without pushing things like the `_TaskState` tracking and management into the submitter abstractions, which would have made the separation less clean IMO. After dynamic block splitting support is added for actors, these two paths should converge.